File: app/src/main/java/sk/styk/martin/apkanalyzer/business/service/PermissionsService.java
package sk.styk.martin.apkanalyzer.business.service;

import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.PermissionInfo;
import android.os.Build;
import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import sk.styk.martin.apkanalyzer.model.detail.PermissionData;
import sk.styk.martin.apkanalyzer.model.detail.PermissionDataAgregate;
import sk.styk.martin.apkanalyzer.model.detail.UsedPermissionData;

/**
 * Created by Martin Styk on 30.06.2017.
 */
public class PermissionsService {

    public PermissionDataAgregate get(@NonNull PackageInfo packageInfo, @NonNull PackageManager packageManager) {

        List<PermissionData> definedPermissions = getDefinedPermissions(packageInfo, packageManager);
        List<UsedPermissionData> requestedPermissions = getUsedPermissions(packageInfo, packageManager);

        return new PermissionDataAgregate(definedPermissions, requestedPermissions);
    }

    public List<PermissionData> getDefinedPermissions(@NonNull PackageInfo packageInfo, @NonNull PackageManager packageManager) {
        PermissionInfo[] permissionInfos = packageInfo.permissions;

        List<PermissionData> permissionData;

        if (permissionInfos == null || permissionInfos.length == 0) {
            permissionData = new ArrayList<>(0);
        } else {
            permissionData = new ArrayList<>(permissionInfos.length);
            for (PermissionInfo permissionInfo : permissionInfos) {
                permissionData.add(getPermissionData(permissionInfo, packageManager));
            }
        }

        return permissionData;
    }

    private PermissionData getPermissionData(@NonNull PermissionInfo permissionInfo, @NonNull PackageManager packageManager) {

        String description = permissionInfo.loadDescription(packageManager) == null ? "" : permissionInfo.loadDescription(packageManager).toString();

        return new PermissionData(
                permissionInfo.name,
                permissionInfo.group,
                description,
                permissionInfo.protectionLevel,
                null
//                permissionInfo.loadIcon(packageManager)
        );

    }

    public List<UsedPermissionData> getUsedPermissions(@NonNull PackageInfo packageInfo, @NonNull PackageManager packageManager) {
        List<UsedPermissionData> requestedPermissions;

        String[] requestedPermissionNames = packageInfo.requestedPermissions;

        int[] requestedPermissionFlags = new int[0];
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
            requestedPermissionFlags = packageInfo.requestedPermissionsFlags;
        }

        if (requestedPermissionNames == null) {
            requestedPermissions = new ArrayList<>(0);
        } else {
            requestedPermissions = new ArrayList<>(requestedPermissionNames.length);

            for (int i = 0; i < requestedPermissionNames.length; i++) {
                boolean isGranted = false;

                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                    isGranted = (requestedPermissionFlags[i] & PackageInfo.REQUESTED_PERMISSION_GRANTED) != 0;
                }
                String name = requestedPermissionNames[i];

                PermissionData permissionData = null;
                try {
                    PermissionInfo permissionInfo = packageManager.getPermissionInfo(name, PackageManager.GET_META_DATA);
                    permissionData = getPermissionData(permissionInfo, packageManager);
                } catch (Exception e) {
                    // we failed to get permission data from pacakge manager. Try to use things we know
                    permissionData = new PermissionData(name, null, null, Integer.MIN_VALUE, null);
                }

                requestedPermissions.add(new UsedPermissionData(permissionData, isGranted));
            }
        }

        return requestedPermissions;
    }

}

File: app/src/main/java/sk/styk/martin/apkanalyzer/business/service/PermissionsService.kt
package sk.styk.martin.apkanalyzer.business.service

import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.os.Build
import sk.styk.martin.apkanalyzer.model.detail.PermissionData
import sk.styk.martin.apkanalyzer.model.detail.PermissionDataAgregate
import sk.styk.martin.apkanalyzer.model.detail.UsedPermissionData

/**
 * @author Martin Styk
 * @version 30.06.2017.
 */
class PermissionsService {

    fun get(packageInfo: PackageInfo, packageManager: PackageManager): PermissionDataAgregate {

        val definedPermissions = getDefinedPermissions(packageInfo)
        val requestedPermissions = getUsedPermissions(packageInfo, packageManager)

        return PermissionDataAgregate(definedPermissions, requestedPermissions)
    }

    private fun getDefinedPermissions(packageInfo: PackageInfo): List<PermissionData> {
        val permissionInfos = packageInfo.permissions ?: return emptyList()

        return permissionInfos.mapTo(ArrayList<PermissionData>(permissionInfos.size)) {
            PermissionData(it.name, it.group, it.protectionLevel)
        }
    }

    fun getUsedPermissions(packageInfo: PackageInfo, packageManager: PackageManager): List<UsedPermissionData> {

        val requestedPermissionNames = packageInfo.requestedPermissions ?: return emptyList()
        val requestedPermissionFlags =
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN)
                    packageInfo.requestedPermissionsFlags
                else
                    IntArray(0)


        val requestedPermissions: MutableList<UsedPermissionData> = ArrayList(requestedPermissionNames.size)

        requestedPermissionNames.indices.forEach { index ->

            val name = requestedPermissionNames[index]
            val isGranted = (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) && (requestedPermissionFlags[index] and PackageInfo.REQUESTED_PERMISSION_GRANTED != 0)

            val permissionData =
                    try {
                        val permissionInfo = packageManager.getPermissionInfo(name, PackageManager.GET_META_DATA)
                        PermissionData(name, permissionInfo.group, permissionInfo.protectionLevel)
                    } catch (e: Exception) {
                        // we failed to get permission data from package manager. Try to use things we know
                        PermissionData(name, null, Integer.MIN_VALUE)
                    }

            requestedPermissions.add(UsedPermissionData(permissionData, isGranted))
        }
        return requestedPermissions

    }

}


